Make GameplayClockContainer abstract and add MasterGameplayClockContainer

smoogipoo opened the pull request:
Prereqs:
- [x] https://github.com/ppy/osu/pull/12496

For multiplayer spectate, I intend to have one master `GameplayClockContainer` controlling the overall timing and track state, and one slave `GameplayClockContainer` for each player instance that sources its time from the master clock.

There should be no logical changes with this code, just abstraction and providing some way to inject the offset clocks into the clock hierarchy of `MasterGameplayClockContainer`

smoogipoo committed:
Abstractify GameplayClockContainer

smoogipoo committed:
Fix IsPaused not being bound

smoogipoo committed:
Fix incorrect offset due to another intermediate Decoupleable clock

smoogipoo committed:
Split out files

smoogipoo committed:
Update clock from base class

smoogipoo committed:
Merge branch 'master' into gcc-abstraction

smoogipoo committed:
Remove ClockToProcess, always process underlying clock

smoogipoo committed:
More refactoring/xmldocs

smoogipoo committed:
Add xmldoc to MasterGameplayClockContainer

smoogipoo committed:
Rename test scene to match class

bdach commented:
Pretty dreadful diff to go through, this. It does look mostly correct to me aside from the above discrepancies, which I pointed out since the test failures seem potentially relevant. I haven't looked too much at them, though, but I can say for sure they seem to be non-deterministic (I didn't get the exact same failure, but other tests from `TestSceneLeadIn` seem to fail interchangeably in the same manner).

smoogipoo commented on osu.Game/Screens/Play/GameplayClockContainer.cs:
```diff
@@ -1,300 +1,126 @@
 
-            adjustableClock = new DecoupleableInterpolatingFramedClock { IsCoupled = false };
+            AdjustableSource = new DecoupleableInterpolatingFramedClock { IsCoupled = false };
+            AdjustableSource.ChangeSource(sourceClock);
```
> Is this just done eagerly here instead on restart as before

Yep, that's exactly it.

smoogipoo replied:
I've moved it back into `Restart()` (now named `Reset()`), because I noticed that changing the source in the ctor causes the track to be restarted in `PlayerLoader`,

smoogipoo committed:
Rename Restart() -> Reset()

smoogipoo committed:
Only set initial source in Reset()

smoogipoo commented on osu.Game/Screens/Play/GameplayClockContainer.cs:
```diff
@@ -1,300 +1,126 @@
-            adjustableClock.ChangeSource(track);
+            if (!IsPaused.Value)
+                Start();
         }
```
Right, I forgot to check this.

- I looked back through blames and found no good reasoning for why this is in a `Task.Run()`. I think it was probably just performance, but it's a few milliseconds at worst, and it's not like everything here is async'd.
- ~~The source change is done in the ctor as above.~~ (see above discussion)
- Rate update is not scheduled because this is completely running on the update thread.

bdach replied:
Removing this schedule broke the reset procedure [as I described on discord](https://discord.com/channels/188630481301012481/188630652340404224/834839320157224961). I am not yet 100% unsure why, and am looking to submit a fix for this, but am leaving this comment in case I fail to do so in the time I have left.

smoogipoo committed:
Fix Reset() potentially not resetting to the intended start position

smoogipoo committed:
Move source clock adjustment application to Start()

smoogipoo committed:
Ensure clock is set in GCC.Start()

smoogipoo committed:
Remove unused usings

smoogipoo committed:
Merge branch 'fix-pre-seek-sample-lifetime' into gcc-abstraction

smoogipoo committed:
Fix compile error

smoogipoo committed:
Move clock processing to base.Seek()

smoogipoo committed:
Seek using local method

smoogipoo committed:
Move seekOffset back to MasterGameplayClockContainer

smoogipoo committed:
Fix storyboard sample test scene

smoogipoo committed:
Fix tests failing intermittently

This was due to this code happening in UpdateAfterChildren(), after the
GCC has processed one frame. During this time, the clock could have
advanced an arbitrary amount.

The cause of this is the removal of the Task.Run() to set the clock in
Restart() (now called Reset()) which changed the timing, so it only worked before due to pure
luck.

smoogipoo commented:
Made a few more changes to solidify logic, before fixing the failing tests in the last commit.

peppy commented on osu.Game/Screens/Play/MasterGameplayClockContainer.cs:
```diff
@@ -0,0 +1,233 @@
+            }
+        }
+
+        private class LocalGameplayClock : GameplayClock
```
Wondering if this should be called `MasterGameplayClock` instead, now that it exists in this class? the "Local" term had me a bit confused at first.

peppy reviewed (commented):
Apart from the mentioned naming this looks solid and an improvement in readability.

smoogipoo committed:
Rename to MasterGameplayClock

smoogipoo committed:
Merge branch 'master' into gcc-abstraction
